docs/proposals/vip-4.md

```diff
@@ -0,0 +1,217 @@
+  * Performing custom health checks on the client applications to ensure that their key metrics are still healthy within
+    a region where a new store-version was swapped, before proceeding to more regions. Based on the outcome of this 
+    check:
+    * Having the ability to abort the swapping of a store-version to further regions.
+    * Having the ability to rollback to the previous store-version in regions that already swapped.
+* Trigger informational events in proprietary monitoring systems after important lifecycle milestones are completed.
+
+## Project Justification
+
+The cost of having global impact in the case of issues mentioned above is too high, and we would like to provide 
+first-class options to reduce the blast radius. Building this within Venice itself will make it easier to automate these
+methodologies, thus reducing toil for users.
+
+## Functional specification
+
+The proposed API for this functionality is described in code here:
+
+[//]: # (Got to remove the /venice prefix before pushing to main on linkedin/venice, otherwise links will be broken...)
+
+* [StoreLifecycleHooks](/venice/javadoc/com/linkedin/venice/hooks/StoreLifecycleHooks.html), which is the main part of this 
+  proposal.
+* [StoreLifecycleEventOutcome](/venice/javadoc/com/linkedin/venice/hooks/StoreLifecycleEventOutcome.html), which is the signal
+  returned by some hooks to indicate that a given step should proceed or abort.
+* [StoreVersionLifecycleEventOutcome](/venice/javadoc/com/linkedin/venice/hooks/StoreVersionLifecycleEventOutcome.html), which
+  is the signal returned by some other hooks which need more fine-grained control over the workflow. In addition to 
+  proceeding and aborting, this also provides the option to wait, which tells the hooks framework to try invoking the 
+  hook again later, and rollback, which tells the framework to rollback to the previous store-version in all regions.
+
+## Proposed Design
+
+The main design consideration is where to execute the hooks. At a high level, there are three options:
+
+1. Within the push job.
+2. Within the parent controller.
+3. Within the child controllers.
+
+It is also possible to consider invoking some hooks in one of these location while other hooks would be executed 
+elsewhere. The table below summarizes the feasibility and tradeoffs for each of the proposed hooks:
+
+|                                              | Child Controller | Parent Controller | Push Job |
+|---------------------------------------------:|:----------------:|:-----------------:|:--------:|
+|                          `preStartOfPushJob` |        ❌         |         ✅         |    ✅     |
+|                         `postStartOfPushJob` |        ❌         |         ✅         |    ✅     |
+|                      `preSchemaRegistration` |        ❌         |         ✅         |   1️⃣    |
+|                     `postSchemaRegistration` |        ✅         |         ✅         |   1️⃣    |
```

Comment:
Schema registration is a global change today, so are you considering to make it region specific to avoid global outage? Also schema rollback is not supported today, and it needs client to support read-on-write schema support.

Reply:
Please take a look at the code included in this PR, which I think makes things more clear than the text.

Some hooks return the `StoreLifecycleEventOutcome` enum, which allows only `PROCEED` or `ABORT`. This is the case of the schema registration hook.

Other hooks return the `StoreVercionLifecycleEventOutcome` enum, which allows not only `PROCEED` and `ABORT`, but also `WAIT` and `ROLLBACK`.

Finally, some hooks return `void` and thus have no ability to affect anything.

IOW, the proposed API uses strong typing to make it very clear and strict what each hook can or cannot accomplish.

Reply:
No... as I said, please take a look at the abstract class in this PR, which should make the contract of each hook clear.

The schema registration has two hooks: pre- and post-registration. Only the pre-registration hook has the ability to abort. The post-hook is `void` and thus for informational purpose only.

Reply:
given we would have schema deletion API, if there a risk in deleting bad as a rollback step? I assume some clients might have already refreshed to the latest schema, but given its bad to beging with, might be ok to delete?

Reply:
Yeah I think it's conceivable to support schema deletion as a hook at some point, though we would need more than just rollback to make it safe... we would need to rollback to the backup version, and wait for the future version we rolled back from to be fully deleted...

Given that we're going in the direction of automatically deleting unused schemas, I don't think we need to involve hooks... we can just let the automation run its course...

Reply:
FYI, I've added a column in the table to indicate the actions each hook is allowed to return, so that the doc is self-contained and reading the code is less important... I hope that helps make things more clear!